File: restaurant/model/order.py
from datetime import datetime
from model.enums import OrderStatus
from model.order_item import OrderItem

class Order:
    """Model đại diện cho một đơn hàng"""
    
    def __init__(self, order_id: int, table_id: int):
        self.order_id = order_id
        self.table_id = table_id
        self.status = OrderStatus.PENDING
        self.items: list[OrderItem] = []
        self.total_amount = 0.0
        self.created_at = datetime.now()
        self.completed_at = None
    
    def add_item(self, order_item: OrderItem):
        """Thêm mục vào đơn hàng"""
        self.items.append(order_item)
        self._calculate_total()
    
    def remove_item(self, order_item_id: int):
        """Xóa mục khỏi đơn hàng"""
        self.items = [item for item in self.items if item.order_item_id != order_item_id]
        self._calculate_total()
    
    def _calculate_total(self):
        """Tính tổng tiền của đơn hàng"""
        self.total_amount = sum(item.get_total_price() for item in self.items)
    
    def change_status(self, new_status: OrderStatus):
        """Thay đổi trạng thái đơn hàng"""
        self.status = new_status
        if new_status == OrderStatus.SERVED:
            self.completed_at = datetime.now()
    
    def __repr__(self):
        return f"Order(id={self.order_id}, table={self.table_id}, status={self.status.value}, total={self.total_amount})"

File: backend/src/com/restaurant/model/entity/Order.java
package com.restaurant.model.entity;

import com.restaurant.model.enums.OrderStatus;
import java.time.LocalDateTime;
import java.util.ArrayList;
import java.util.List;

/**
 * Entity đại diện cho một đơn hàng
 */
public class Order {
    private final int orderId;
    private final int tableId;
    private OrderStatus status;
    private final List<OrderItem> items;
    private double totalAmount;
    private final LocalDateTime createdAt;
    private LocalDateTime completedAt;

    public Order(int orderId, int tableId) {
        this.orderId = orderId;
        this.tableId = tableId;
        this.status = OrderStatus.NEW;
        this.items = new ArrayList<>();
        this.totalAmount = 0.0;
        this.createdAt = LocalDateTime.now();
        this.completedAt = null;
    }

    // Getters
    public int getOrderId() {
        return orderId;
    }

    public int getTableId() {
        return tableId;
    }

    public OrderStatus getStatus() {
        return status;
    }

    public List<OrderItem> getItems() {
        return new ArrayList<>(items);
    }

    public double getTotalAmount() {
        return totalAmount;
    }

    public LocalDateTime getCreatedAt() {
        return createdAt;
    }

    public LocalDateTime getCompletedAt() {
        return completedAt;
    }

    // Business Methods
    public void addItem(OrderItem orderItem) {
        items.add(orderItem);
        calculateTotal();
    }

    public boolean removeItem(int orderItemId) {
        boolean removed = items.removeIf(item -> item.getOrderItemId() == orderItemId);
        if (removed) {
            calculateTotal();
        }
        return removed;
    }

    private void calculateTotal() {
        totalAmount = items.stream()
                .mapToDouble(OrderItem::getTotalPrice)
                .sum();
    }

    public void changeStatus(OrderStatus newStatus) {
        this.status = newStatus;
        if (newStatus == OrderStatus.COMPLETED) {
            this.completedAt = LocalDateTime.now();
        }
    }

    public int getItemCount() {
        return items.size();
    }

    public boolean isEmpty() {
        return items.isEmpty();
    }

    @Override
    public String toString() {
        return String.format("Đơn hàng #%d - Bàn %d - Tổng: %,.0fđ",
                orderId, tableId, totalAmount);
    }
}
